src/BinSkim.Driver/RuleTelemetryLogger.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.CodeAnalysis.Sarif;
+
+namespace Microsoft.CodeAnalysis.IL
+{
+    /// <summary>
+    /// A logger that records rule summary telemetry by aggregating result kinds for each rule.
+    /// </summary>
+    internal sealed class RuleTelemetryLogger : IAnalysisLogger
+    {
+        // Application Insights event names
+        internal const string RuleSummaryEventName = "RuleSummary";
+        internal const string AnalysisStoppedEventName = "AnalysisStopped";
+
+        // Application Insights property names
+        internal const string RuleIdPropertyName = "RuleId";
+        internal const string RuntimeConditionsPropertyName = "Flags";
+
+        // Application Insights metric names
+        internal const string PassCountMetricName = "Pass";
+        internal const string FailCountMetricName = "Fail";
+        internal const string OpenCountMetricName = "Open";
+        internal const string ReviewCountMetricName = "Review";
+        internal const string NotApplicableCountMetricName = "NA";
+        internal const string InformationalCountMetricName = "Info";
+
+        /// <summary>
+        /// The Application Insights telemetry client.
+        /// </summary>
+        private readonly TelemetryClient telemetryClient;
+
+        /// <summary>
+        /// Dictionary of aggregated counts of result kinds keyed on the rule ID.
+        /// </summary>
+        private readonly Dictionary<string, ResultKindCounts> metricsMap = new Dictionary<string, ResultKindCounts>(StringComparer.Ordinal);
+
+        /// <summary>
+        /// Construct a new <see cref="RuleTelemetryLogger"/>.
+        /// </summary>
+        /// <param name="telemetryClient">The Application Insights telemetry client.</param>
+        public RuleTelemetryLogger(TelemetryClient telemetryClient)
+        {
+            this.telemetryClient = telemetryClient ?? throw new ArgumentNullException(nameof(telemetryClient));
+        }
+
+        public void AnalysisStarted()
+        {
+        }
+
+        public void AnalysisStopped(RuntimeConditions runtimeConditions)
+        {
+            EventTelemetry stopEvent = new EventTelemetry(AnalysisStoppedEventName);
```

Comment:
>EventTelemetry [](http://example.com/codeflow?start=12&length=14)

generally we prefer to use 'var' when assigning object construction with clearly specified type. we're supposed to be configured to produce warnings in the IDE on this, might not be properly enabled/flowing to you (or maybe you're using vs code not vs).